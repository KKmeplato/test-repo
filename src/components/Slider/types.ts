import * as RadixSlider from "@radix-ui/react-slider";

export interface SliderProps
  extends Omit<
    RadixSlider.SliderProps,
    "onValueChange" | "onChange" | "onValueCommit" | "asChild" | "orientation"
  > {
  onChange?: RadixSlider.SliderProps["onValueChange"];
  onChangeCommit?: RadixSlider.SliderProps["onValueCommit"];
  prefix?: string;
  suffix?: string;
  label?: string;
  id?: string;
  accessibilityLabel?: string;
}

export interface SliderThumbProps extends RadixSlider.SliderThumbProps {
  forceTooltipOpen: boolean;
  onOpenChange: (open: boolean) => void;
  value: number;
}
