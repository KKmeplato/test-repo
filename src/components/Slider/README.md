### Slider

```tsx
<Slider
  label=""
  defaultValue={[10]}
  value={[10]}
  min={0}
  max={100}
  step={1}
  disabled={false}
  onChange={(value) => {}}
  onChangeCommit={(value) => {}}
/>
```

#### Range slider

```tsx
<Slider
  label=""
  defaultValue={[10, 60]}
  value={[10, 60]}
  min={0}
  max={100}
  step={1}
  disabled={false}
  onChange={(value) => {}}
  onChangeCommit={(value) => {}}
/>
```

- [x] Meplato light theme
- [x] Meplato dark theme
- [x] Webridge light theme
- [x] Webridge dark theme
