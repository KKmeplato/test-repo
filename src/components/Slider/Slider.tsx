import * as RadixSlider from "@radix-ui/react-slider";
import React, { useId, useMemo, useState } from "react";

import { cn } from "lib/utils";

import { SliderThumb } from "./components";
import { SliderProps } from "./types";

export const Slider = React.forwardRef<HTMLDivElement, SliderProps>(
  (
    {
      value,
      prefix,
      suffix,
      label,
      id: _id,
      className,
      disabled,
      accessibilityLabel,
      onChange,
      onChangeCommit,
      defaultValue,
      ...rest
    },
    forwardedRef
  ) => {
    const prefixSuffixClassName = "text-sm";
    const generatedId = useId();
    const [isSliderHovered, setSliderHovered] = useState(false);
    const id = _id || generatedId;
    const [thumbs, setThumbs] = useState(
      (value || defaultValue)?.map(() => ({
        tooltipOpen: false
      })) || []
    );

    const isThumbTooltipOpen = useMemo(() => {
      return Boolean(
        thumbs.filter(({ tooltipOpen }) => Boolean(tooltipOpen))?.length
      );
    }, [thumbs]);

    return (
      <div
        className={cn("group flex flex-col gap-1", {
          "text-disabled": disabled
        })}
      >
        {label && (
          <label className="text-sm" id={id}>
            {label}
          </label>
        )}
        <div className="flex items-center gap-2">
          {prefix && <span className={prefixSuffixClassName}>{prefix}</span>}
          <RadixSlider.Root
            {...rest}
            orientation="horizontal"
            defaultValue={defaultValue}
            disabled={disabled}
            ref={forwardedRef}
            className={cn(
              "relative flex w-full touch-none select-none items-center py-3",
              { "pointer-events-none": disabled },
              className
            )}
            onValueCommit={onChangeCommit}
            onValueChange={onChange}
            value={value}
            onMouseEnter={() => {
              setSliderHovered(true);
            }}
            onMouseLeave={() => {
              setSliderHovered(false);
            }}
          >
            <RadixSlider.Track className="relative grow rounded-full border-[2px] border-dashed border-surface-disabled data-[disabled]:border-solid">
              <RadixSlider.Range className="absolute ml-[-2px] mt-[-2px] h-full rounded-full border-[2px] border-interactive-primary data-[disabled]:border-disabled group-hover:[&:not([data-disabled])]:border-interactive-primary-hovered group-active:[&:not([data-disabled])]:border-interactive-primary" />
            </RadixSlider.Track>
            {(value || defaultValue)?.map((v, index) => (
              <SliderThumb
                key={index}
                aria-labelledby={(label && id) || undefined}
                aria-label={accessibilityLabel}
                forceTooltipOpen={isThumbTooltipOpen || isSliderHovered}
                onOpenChange={(open) => {
                  const updatedThumbs = [...thumbs];

                  if (!updatedThumbs?.[index]) {
                    return;
                  }

                  if (updatedThumbs[index]["tooltipOpen"] === open) {
                    return;
                  }
                  updatedThumbs[index] = {
                    ...updatedThumbs[index],
                    tooltipOpen: open
                  };
                  setThumbs(updatedThumbs);
                }}
                value={v}
              />
            ))}
          </RadixSlider.Root>
          {suffix && <span className={prefixSuffixClassName}>{suffix}</span>}
        </div>
      </div>
    );
  }
);

Slider.displayName = "Slider";
