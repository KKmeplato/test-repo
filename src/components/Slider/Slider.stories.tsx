import { action } from "@storybook/addon-actions";
import { Meta, StoryObj } from "@storybook/react";
import React, { useState } from "react";

import { Slider } from "./Slider";
import { SliderProps } from "./types";

const meta: Meta<typeof Slider> = {
  component: Slider,
  decorators: [
    (Story) => {
      return (
        <div className="max-w-sm py-16">
          <Story />
        </div>
      );
    }
  ]
};

export default meta;
type Story = StoryObj<typeof Slider>;

function SliderWithArgs(args: SliderProps) {
  const [value, setValue] = useState(args.defaultValue);
  const handleChange = React.useCallback((value: number[]) => {
    setValue(value);
  }, []);

  return (
    <Slider
      {...args}
      value={value}
      onChange={handleChange}
      onChangeCommit={action("onChangeCommit")}
    />
  );
}

const createStory = ({
  defaultValue
}: {
  defaultValue: SliderProps["defaultValue"];
}): Story => ({
  args: {
    defaultValue,
    step: 1,
    min: 0,
    max: 100,
    disabled: false,
    prefix: "$",
    suffix: "100",
    label: "Label",
    accessibilityLabel: defaultValue?.length === 1 ? "Price" : "Price range"
  },
  render: SliderWithArgs
});

export const SingleThumb = createStory({
  defaultValue: [40]
});

export const DualThumb = createStory({
  defaultValue: [40, 100]
});
