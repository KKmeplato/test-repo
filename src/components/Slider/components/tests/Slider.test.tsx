import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Slider } from "../../Slider";

describe("<Slider />", () => {
  it.each([
    {
      name: "single",
      defaultValue: [10],
      value: [10],
      min: 0,
      max: 100,
      label: "Single",
      accessibilityLabel: "Volume"
    },
    {
      name: "range",
      defaultValue: [10, 50],
      value: [10, 50],
      min: 0,
      max: 100,
      label: "Range",
      accessibilityLabel: "Price range"
    }
  ])("renders with $name", (args) => {
    const result = render(<Slider {...args} onChange={vi.fn()} />);

    expect(result).toMatchSnapshot();
  });

  it("renders a slider based on props", () => {
    const { getByRole, getByLabelText } = render(
      <Slider
        defaultValue={[10]}
        min={0}
        max={100}
        value={[10]}
        label="Volume"
        prefix=""
        onChange={vi.fn()}
      />
    );

    expect(
      getByRole("slider", { value: { now: 10, min: 0, max: 100 } })
    ).toBeInTheDocument();
    expect(getByLabelText("Volume")).toBeInTheDocument();
  });

  it("renders a range slider based on props", () => {
    const { getByRole, getAllByLabelText } = render(
      <Slider
        defaultValue={[10]}
        min={0}
        max={100}
        value={[10, 60]}
        label="Price range"
        prefix=""
        onChange={vi.fn()}
      />
    );

    expect(
      getByRole("slider", { value: { now: 10, min: 0, max: 100 } })
    ).toBeInTheDocument();
    expect(
      getByRole("slider", { value: { now: 60, min: 0, max: 100 } })
    ).toBeInTheDocument();
    expect(getAllByLabelText("Price range").length).toBe(2);
  });

  it("shows a tooltip when hovered on the slider thumb", async () => {
    const { getByRole } = render(
      <Slider
        defaultValue={[10]}
        min={0}
        max={100}
        value={[10]}
        label="Volume"
        prefix=""
        onChange={vi.fn()}
      />
    );

    await userEvent.hover(getByRole("slider"));
    expect(getByRole("tooltip")).toHaveTextContent("10");
  });

  it("shows a tooltip when hovered on the slider thumb when dual thumbs are used", async () => {
    const { getAllByRole } = render(
      <Slider
        defaultValue={[10]}
        min={0}
        max={100}
        value={[10, 60]}
        label="Price range"
        prefix=""
        onChange={vi.fn()}
      />
    );

    await userEvent.hover(getAllByRole("slider")[0]);
    expect(getAllByRole("tooltip")[0]).toHaveTextContent("10");

    await userEvent.hover(getAllByRole("slider")[1]);
    expect(getAllByRole("tooltip")[1]).toHaveTextContent("60");
  });
});
