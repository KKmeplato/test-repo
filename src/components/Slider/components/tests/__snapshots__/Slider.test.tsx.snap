// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Slider /> > renders with 'range' 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="group flex flex-col gap-1"
      >
        <label
          class="text-sm"
          id=":r2:"
        >
          Range
        </label>
        <div
          class="flex items-center gap-2"
        >
          <span
            aria-disabled="false"
            class="relative flex w-full touch-none select-none items-center py-3"
            data-orientation="horizontal"
            dir="ltr"
            style="--radix-slider-thumb-transform: translateX(-50%);"
          >
            <span
              class="relative grow rounded-full border-[2px] border-dashed border-surface-disabled data-[disabled]:border-solid"
              data-orientation="horizontal"
            >
              <span
                class="absolute ml-[-2px] mt-[-2px] h-full rounded-full border-[2px] border-interactive-primary data-[disabled]:border-disabled group-hover:[&:not([data-disabled])]:border-interactive-primary-hovered group-active:[&:not([data-disabled])]:border-interactive-primary"
                data-orientation="horizontal"
                style="left: 10%; right: 50%;"
              />
            </span>
            <span
              style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(10% + 0px);"
            >
              <span
                aria-label="Price range"
                aria-labelledby=":r2:"
                aria-orientation="horizontal"
                aria-valuemax="100"
                aria-valuemin="0"
                aria-valuenow="10"
                class="block h-4 w-4 rounded-[10px] bg-interactive-primary focus-visible:outline focus-visible:outline-1 focus-visible:outline-focus focus-visible:ring-1 focus-visible:ring-focus focus-visible:ring-offset-1 group-active:focus-visible:scale-[1.35] data-[disabled]:bg-interactive-disabled group-hover:[&:not([data-disabled])]:bg-interactive-primary-hovered group-active:[&:not([data-disabled])]:bg-interactive-primary"
                data-orientation="horizontal"
                data-radix-collection-item=""
                role="slider"
                style=""
                tabindex="0"
              />
            </span>
            <span
              style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(50% + 0px);"
            >
              <span
                aria-label="Price range"
                aria-labelledby=":r2:"
                aria-orientation="horizontal"
                aria-valuemax="100"
                aria-valuemin="0"
                aria-valuenow="50"
                class="block h-4 w-4 rounded-[10px] bg-interactive-primary focus-visible:outline focus-visible:outline-1 focus-visible:outline-focus focus-visible:ring-1 focus-visible:ring-focus focus-visible:ring-offset-1 group-active:focus-visible:scale-[1.35] data-[disabled]:bg-interactive-disabled group-hover:[&:not([data-disabled])]:bg-interactive-primary-hovered group-active:[&:not([data-disabled])]:bg-interactive-primary"
                data-orientation="horizontal"
                data-radix-collection-item=""
                role="slider"
                style=""
                tabindex="0"
              />
            </span>
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="group flex flex-col gap-1"
    >
      <label
        class="text-sm"
        id=":r2:"
      >
        Range
      </label>
      <div
        class="flex items-center gap-2"
      >
        <span
          aria-disabled="false"
          class="relative flex w-full touch-none select-none items-center py-3"
          data-orientation="horizontal"
          dir="ltr"
          style="--radix-slider-thumb-transform: translateX(-50%);"
        >
          <span
            class="relative grow rounded-full border-[2px] border-dashed border-surface-disabled data-[disabled]:border-solid"
            data-orientation="horizontal"
          >
            <span
              class="absolute ml-[-2px] mt-[-2px] h-full rounded-full border-[2px] border-interactive-primary data-[disabled]:border-disabled group-hover:[&:not([data-disabled])]:border-interactive-primary-hovered group-active:[&:not([data-disabled])]:border-interactive-primary"
              data-orientation="horizontal"
              style="left: 10%; right: 50%;"
            />
          </span>
          <span
            style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(10% + 0px);"
          >
            <span
              aria-label="Price range"
              aria-labelledby=":r2:"
              aria-orientation="horizontal"
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="10"
              class="block h-4 w-4 rounded-[10px] bg-interactive-primary focus-visible:outline focus-visible:outline-1 focus-visible:outline-focus focus-visible:ring-1 focus-visible:ring-focus focus-visible:ring-offset-1 group-active:focus-visible:scale-[1.35] data-[disabled]:bg-interactive-disabled group-hover:[&:not([data-disabled])]:bg-interactive-primary-hovered group-active:[&:not([data-disabled])]:bg-interactive-primary"
              data-orientation="horizontal"
              data-radix-collection-item=""
              role="slider"
              style=""
              tabindex="0"
            />
          </span>
          <span
            style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(50% + 0px);"
          >
            <span
              aria-label="Price range"
              aria-labelledby=":r2:"
              aria-orientation="horizontal"
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="50"
              class="block h-4 w-4 rounded-[10px] bg-interactive-primary focus-visible:outline focus-visible:outline-1 focus-visible:outline-focus focus-visible:ring-1 focus-visible:ring-focus focus-visible:ring-offset-1 group-active:focus-visible:scale-[1.35] data-[disabled]:bg-interactive-disabled group-hover:[&:not([data-disabled])]:bg-interactive-primary-hovered group-active:[&:not([data-disabled])]:bg-interactive-primary"
              data-orientation="horizontal"
              data-radix-collection-item=""
              role="slider"
              style=""
              tabindex="0"
            />
          </span>
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Slider /> > renders with 'single' 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="group flex flex-col gap-1"
      >
        <label
          class="text-sm"
          id=":r0:"
        >
          Single
        </label>
        <div
          class="flex items-center gap-2"
        >
          <span
            aria-disabled="false"
            class="relative flex w-full touch-none select-none items-center py-3"
            data-orientation="horizontal"
            dir="ltr"
            style="--radix-slider-thumb-transform: translateX(-50%);"
          >
            <span
              class="relative grow rounded-full border-[2px] border-dashed border-surface-disabled data-[disabled]:border-solid"
              data-orientation="horizontal"
            >
              <span
                class="absolute ml-[-2px] mt-[-2px] h-full rounded-full border-[2px] border-interactive-primary data-[disabled]:border-disabled group-hover:[&:not([data-disabled])]:border-interactive-primary-hovered group-active:[&:not([data-disabled])]:border-interactive-primary"
                data-orientation="horizontal"
                style="left: 0%; right: 90%;"
              />
            </span>
            <span
              style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(10% + 0px);"
            >
              <span
                aria-label="Volume"
                aria-labelledby=":r0:"
                aria-orientation="horizontal"
                aria-valuemax="100"
                aria-valuemin="0"
                aria-valuenow="10"
                class="block h-4 w-4 rounded-[10px] bg-interactive-primary focus-visible:outline focus-visible:outline-1 focus-visible:outline-focus focus-visible:ring-1 focus-visible:ring-focus focus-visible:ring-offset-1 group-active:focus-visible:scale-[1.35] data-[disabled]:bg-interactive-disabled group-hover:[&:not([data-disabled])]:bg-interactive-primary-hovered group-active:[&:not([data-disabled])]:bg-interactive-primary"
                data-orientation="horizontal"
                data-radix-collection-item=""
                role="slider"
                style=""
                tabindex="0"
              />
            </span>
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="group flex flex-col gap-1"
    >
      <label
        class="text-sm"
        id=":r0:"
      >
        Single
      </label>
      <div
        class="flex items-center gap-2"
      >
        <span
          aria-disabled="false"
          class="relative flex w-full touch-none select-none items-center py-3"
          data-orientation="horizontal"
          dir="ltr"
          style="--radix-slider-thumb-transform: translateX(-50%);"
        >
          <span
            class="relative grow rounded-full border-[2px] border-dashed border-surface-disabled data-[disabled]:border-solid"
            data-orientation="horizontal"
          >
            <span
              class="absolute ml-[-2px] mt-[-2px] h-full rounded-full border-[2px] border-interactive-primary data-[disabled]:border-disabled group-hover:[&:not([data-disabled])]:border-interactive-primary-hovered group-active:[&:not([data-disabled])]:border-interactive-primary"
              data-orientation="horizontal"
              style="left: 0%; right: 90%;"
            />
          </span>
          <span
            style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(10% + 0px);"
          >
            <span
              aria-label="Volume"
              aria-labelledby=":r0:"
              aria-orientation="horizontal"
              aria-valuemax="100"
              aria-valuemin="0"
              aria-valuenow="10"
              class="block h-4 w-4 rounded-[10px] bg-interactive-primary focus-visible:outline focus-visible:outline-1 focus-visible:outline-focus focus-visible:ring-1 focus-visible:ring-focus focus-visible:ring-offset-1 group-active:focus-visible:scale-[1.35] data-[disabled]:bg-interactive-disabled group-hover:[&:not([data-disabled])]:bg-interactive-primary-hovered group-active:[&:not([data-disabled])]:bg-interactive-primary"
              data-orientation="horizontal"
              data-radix-collection-item=""
              role="slider"
              style=""
              tabindex="0"
            />
          </span>
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
